Add CodeQL GitHub Action

masterleinad opened the pull request:
In response to https://securityscorecards.dev/viewer/?uri=github.com/kokkos/kokkos and https://kokkosteam.slack.com/archives/G5CBLMFLP/p1707781171873809. Following https://docs.github.com/en/code-security/code-scanning/creating-an-advanced-setup-for-code-scanning/configuring-advanced-setup-for-code-scanning.

crtrott commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,53 @@
+name: "CodeQL"
+
```
we should add the permissions thing I added in the other workflows.

masterleinad replied:
What we have is more restrictive already in most cases and we need to be able to write to security events.

crtrott replied:
THe OpenSSF Scorecard I think wants a default premission you then can overwrite.

crtrott commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,53 @@
+  push:
+    branches: [ "master", "develop", "release-*" ]
+  pull_request:
+    branches: [ "master", "develop", "release-*" ]
```
Should we just run on push?

crtrott commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,53 @@
+          -DCMAKE_CXX_STANDARD=17
+          -DKokkos_ENABLE_COMPILER_WARNINGS=ON
+          -DKokkos_ENABLE_DEPRECATED_CODE_4=OFF
+          -DKokkos_ENABLE_TESTS=On
```
Add examples and benchmarks?

crtrott reviewed (changes requested):
what did you customize from the default one?

masterleinad commented:
> what did you customize from the default one?

The default one is
```
# For most projects, this workflow file will not need changing; you simply need
# to commit it to your repository.
#
# You may wish to alter this file to override the set of languages analyzed,
# or to provide custom queries or build logic.
#
# ******** NOTE ********
# We have attempted to detect the languages in your repository. Please check
# the `language` matrix defined below to confirm you have the correct set of
# supported CodeQL languages.
#
name: "CodeQL"

on:
  push:
    branches: [ "master" ]
  pull_request:
    branches: [ "master" ]
  schedule:
    - cron: '24 14 * * 2'

jobs:
  analyze:
    name: Analyze
    # Runner size impacts CodeQL analysis time. To learn more, please see:
    #   - https://gh.io/recommended-hardware-resources-for-running-codeql
    #   - https://gh.io/supported-runners-and-hardware-resources
    #   - https://gh.io/using-larger-runners
    # Consider using larger runners for possible analysis time improvements.
    runs-on: ${{ (matrix.language == 'swift' && 'macos-latest') || 'ubuntu-latest' }}
    timeout-minutes: ${{ (matrix.language == 'swift' && 120) || 360 }}
    permissions:
      # required for all workflows
      security-events: write

      # only required for workflows in private repositories
      actions: read
      contents: read

    strategy:
      fail-fast: false
      matrix:
        language: [ 'c-cpp', 'python' ]
        # CodeQL supports [ 'c-cpp', 'csharp', 'go', 'java-kotlin', 'javascript-typescript', 'python', 'ruby', 'swift' ]
        # Use only 'java-kotlin' to analyze code written in Java, Kotlin or both
        # Use only 'javascript-typescript' to analyze code written in JavaScript, TypeScript or both
        # Learn more about CodeQL language support at https://aka.ms/codeql-docs/language-support

    steps:
    - name: Checkout repository
      uses: actions/checkout@v4

    # Initializes the CodeQL tools for scanning.
    - name: Initialize CodeQL
      uses: github/codeql-action/init@v3
      with:
        languages: ${{ matrix.language }}
        # If you wish to specify custom queries, you can do so here or in a config file.
        # By default, queries listed here will override any specified in a config file.
        # Prefix the list here with "+" to use these queries and those in the config file.

        # For more details on CodeQL's query packs, refer to: https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#using-queries-in-ql-packs
        # queries: security-extended,security-and-quality


    # Autobuild attempts to build any compiled languages (C/C++, C#, Go, Java, or Swift).
    # If this step fails, then you should remove it and run the build manually (see below)
    - name: Autobuild
      uses: github/codeql-action/autobuild@v3

    # ℹ️ Command-line programs to run using the OS shell.
    # 📚 See https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsrun

    #   If the Autobuild fails above, remove it and uncomment the following three lines.
    #   modify them (or add more) to build your code if your project, please refer to the EXAMPLE below for guidance.

    # - run: |
    #     echo "Run, Build Application using script"
    #     ./location_of_script_within_repo/buildscript.sh

    - name: Perform CodeQL Analysis
      uses: github/codeql-action/analyze@v3
      with:
        category: "/language:${{matrix.language}}"
```
so basically I changed
```
    runs-on: ${{ (matrix.language == 'swift' && 'macos-latest') || 'ubuntu-latest' }}
    timeout-minutes: ${{ (matrix.language == 'swift' && 120) || 360 }}
     language: [ 'c-cpp', 'python' ]
```
to 
```
    runs-on:  ubuntu-latest
    timeout-minutes: 360
    language: c-cpp
```
and replaced the `Autobuild` part with a `OpenMP` configuration.

masterleinad commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,55 @@
+on:
+  push:
+    branches: [ "master", "develop", "release-*" ]
+  pull_request:
+    branches: [ "develop" ]
```
I can remove again before merging if we don't want to change anything else.

dalg24 commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,55 @@
+        cmake -B build .
+          -DKokkos_ENABLE_OPENMP=ON
+          -DCMAKE_CXX_STANDARD=17
+          -DKokkos_ENABLE_COMPILER_WARNINGS=ON
```
What are the compiler warnings good for in that build?

masterleinad replied:
Let me drop it.

dalg24 commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,55 @@
+    - name: build
+      run:
+        cmake --build build --parallel 2
+    - name: test
+      working-directory: build
+      run: ctest --output-on-failure
```
I thought it was a static analysis.  What does actually running the tests bring?

masterleinad replied:
Let's just remove it and see how the report differs. Ultimately, we could take the opportunity to test another configuration here but the current one is not very special anyway.

masterleinad replied:
Runtime dropped from 69 min to 48 min. The reports look similar.

dalg24 reviewed (approved):
Let's give it a shot
